File: frontend/src/services/sessionSchedule.ts
export interface SessionSchedule {
  name: string
  date?: string
  admission_number?: string
  patient_num?: string
  patient_visit?: string
  session_type?: string
  session_name?: string
  transaction_status?: string
  company?: string
  doctor?: string
  doctor_name?: string
  practitioner?: string
  practitioner_name?: string
  cost_center?: string
  invoice_no?: string
  doc_code?: string
  from_time?: string
  to_time?: string
  amount?: number
  sales_order?: string
  doc_remarks?: string
  feedback_remarks?: string
  sr_num?: string
}

export interface CreateSessionScheduleData {
  date: string
  /** Patient id — required when visit/admission is not linked */
  patient?: string
  admission_number?: string
  patient_visit?: string
  session_type: string
  session_name?: string
  doctor?: string
  /** Healthcare Practitioner who entered the session. */
  practitioner?: string
  practitioner_name?: string
  cost_center?: string
  from_time?: string
  to_time?: string
  amount?: number
  doc_remarks?: string
}

export async function fetchSessionSchedules(
  limit: number = 50,
  offset: number = 0,
  patient?: string,
  admissionNumber?: string,
  roleGroup?: string,
  practitioner?: string
): Promise<{ data: SessionSchedule[]; total_count: number }> {
  const params = new URLSearchParams()
  params.append('limit', limit.toString())
  params.append('offset', offset.toString())
  if (patient) params.append('patient', patient)
  if (admissionNumber) params.append('admission_number', admissionNumber)
  if (roleGroup) params.append('role_group', roleGroup)
  if (practitioner) params.append('practitioner', practitioner)

  const response = await fetch(
    `/api/method/healthcare.api.session_schedule.get_session_schedules?${params.toString()}`
  )
  const resData = await response.json()
  const message = resData?.message

  if (Array.isArray(message)) {
    return { data: message as SessionSchedule[], total_count: message.length }
  }
  if (message && typeof message === 'object' && Array.isArray(message.data)) {
    return {
      data: message.data as SessionSchedule[],
      total_count: Number(message.total_count ?? message.data.length),
    }
  }
  return { data: [], total_count: 0 }
}

export async function fetchSessionSchedule(name: string): Promise<SessionSchedule> {
  const { fetchDoc } = await import('./common')
  const data = await fetchDoc('Session Schedule', name)
  return { ...(data as unknown as SessionSchedule), name: String(data.name || name) }
}

/** Display name for Session Schedule.practitioner (Username on the detail panel). */
export async function resolveSessionPractitionerName(
  practitioner?: string | null,
  practitionerName?: string | null,
): Promise<string> {
  const labeled = (practitionerName || '').trim()
  if (labeled) return labeled
  const id = (practitioner || '').trim()
  if (!id) return ''
  try {
    const params = new URLSearchParams({
      doctype: 'Healthcare Practitioner',
      fields: JSON.stringify(['name', 'practitioner_name']),
      filters: JSON.stringify([['name', '=', id]]),
      limit_page_length: '1',
    })
    const res = await fetch(`/api/method/frappe.client.get_list?${params}`)
    const payload = await res.json()
    const row = Array.isArray(payload?.message) ? payload.message[0] : null
    const resolved = String(row?.practitioner_name || '').trim()
    if (resolved) return resolved
  } catch {
    /* leave blank rather than showing an internal id */
  }
  return ''
}

export async function createSessionSchedule(data: CreateSessionScheduleData): Promise<SessionSchedule> {
  const csrf = (window as any).csrf_token
  const response = await fetch(
    '/api/method/healthcare.api.session_schedule.create_session_schedule',
    {
      method: 'POST',
      credentials: 'include',
      headers: {
        'Content-Type': 'application/json',
        'Accept': 'application/json',
        ...(csrf ? { 'X-Frappe-CSRF-Token': csrf } : {})
      },
      body: JSON.stringify({ data })
    }
  )
  const resData = await response.json()

  if (resData?.message) {
    return resData.message as SessionSchedule
  } else {
    throw new Error(resData?.exc || 'Failed to create session schedule')
  }
}

export async function updateSessionSchedule(
  name: string,
  data: CreateSessionScheduleData,
): Promise<SessionSchedule> {
  const { apiRequest } = await import('./apiClient')
  return apiRequest<SessionSchedule>(
    '/api/method/healthcare.api.session_schedule.update_session_schedule',
    {
      method: 'POST',
      body: JSON.stringify({
        session_schedule_name: name,
        data,
      }),
    },
  )
}

export async function updateSessionScheduleStatus(
  sessionScheduleName: string,
  status: string,
): Promise<SessionSchedule> {
  const { apiRequest } = await import('./apiClient')
  return apiRequest<SessionSchedule>(
    '/api/method/healthcare.api.session_schedule.update_session_schedule_status',
    {
      method: 'POST',
      body: JSON.stringify({
        session_schedule_name: sessionScheduleName,
        status,
      }),
    },
  )
}

export interface HealthcareServiceTemplateOption {
  name: string
  service_name?: string
  category?: string
  rate?: number
}

export async function getHealthcareServiceTemplates(
  search?: string,
  limit: number = 100,
  patientCareType?: 'OP' | 'IP',
): Promise<HealthcareServiceTemplateOption[]> {
  const params = new URLSearchParams()
  params.set('limit', String(limit))
  if (search?.trim()) params.set('search', search.trim())
  if (patientCareType) params.set('patient_care_type', patientCareType)

  const response = await fetch(
    `/api/method/healthcare.api.ip_service.get_ip_service_types?${params.toString()}`,
  )
  const resData = await response.json()

  if (resData?.message && Array.isArray(resData.message)) {
    return resData.message as HealthcareServiceTemplateOption[]
  }
  return []
}

export interface SessionScheduleAmountPreview {
  /** Inclusive / catalog list price (before insurance %) */
  amount: number
  base_rate?: number
  discount_pct?: number
  discount_amount?: number
  net_rate?: number
  insurance?: string | null
  used_insurance_price?: boolean
  patient?: string | null
  patient_care_type?: string | null
}

/** Insured amount for a Healthcare Service Template on Session Schedule (Inclusive price + OP/IP %). */
export async function fetchSessionScheduleAmount(args: {
  sessionType: string
  patient?: string
  patientVisit?: string
  admissionNumber?: string
  patientCareType?: 'OP' | 'IP'
}): Promise<SessionScheduleAmountPreview> {
  const params = new URLSearchParams()
  params.set('session_type', args.sessionType)
  if (args.patient) params.set('patient', args.patient)
  if (args.patientVisit) params.set('patient_visit', args.patientVisit)
  if (args.admissionNumber) params.set('admission_number', args.admissionNumber)
  if (args.patientCareType) params.set('patient_care_type', args.patientCareType)

  const response = await fetch(
    `/api/method/healthcare.api.session_schedule.get_session_schedule_amount?${params.toString()}`,
    { credentials: 'include' },
  )
  const resData = await response.json()
  const msg = (resData?.message || {}) as Partial<SessionScheduleAmountPreview>
  return {
    amount: Number(msg.amount) || 0,
    base_rate: msg.base_rate != null ? Number(msg.base_rate) : undefined,
    discount_pct: msg.discount_pct != null ? Number(msg.discount_pct) : undefined,
    discount_amount: msg.discount_amount != null ? Number(msg.discount_amount) : undefined,
    net_rate: msg.net_rate != null ? Number(msg.net_rate) : undefined,
    insurance: msg.insurance ?? null,
    used_insurance_price: Boolean(msg.used_insurance_price),
    patient: msg.patient ?? null,
    patient_care_type: msg.patient_care_type ?? null,
  }
}

/** @deprecated Use getHealthcareServiceTemplates */
export async function getSessionTypes(): Promise<HealthcareServiceTemplateOption[]> {
  return getHealthcareServiceTemplates()
}

export async function createSessionScheduleSalesOrder(
  sessionScheduleName: string,
): Promise<{ sales_order: string; existing?: boolean; transaction_status?: string }> {
  const { apiRequest } = await import('./apiClient')
  return apiRequest<{ sales_order: string; existing?: boolean; transaction_status?: string }>(
    '/api/method/healthcare.api.session_schedule.create_sales_order_from_session_schedule',
    {
      method: 'POST',
      body: JSON.stringify({ session_schedule_name: sessionScheduleName }),
    },
  )
}

/** @deprecated Use createSessionScheduleSalesOrder */
export const billSessionSchedule = createSessionScheduleSalesOrder

File: healthcare/api/session_schedule.py
import frappe
from frappe import _
from frappe.utils import cint, flt, getdate, nowdate

from healthcare.healthcare.editing_lock import assert_editing_allowed

SESSION_SCHEDULE_STATUSES = frozenset(
	{"Draft", "In Progress", "Completed", "Submitted", "Cancelled"}
)


def _existing_session_schedule_sales_order(session_schedule_name):
	return frappe.db.get_value(
		"Sales Order",
		{
			"custom_base_reference": "Session Schedule",
			"custom_base_reference_name": session_schedule_name,
			"docstatus": ["!=", 2],
		},
		"name",
	)


def _attach_sales_orders(schedules):
	if not schedules:
		return schedules

	names = [row["name"] for row in schedules if row.get("name")]
	by_schedule = {}
	if names:
		for row in frappe.get_all(
			"Sales Order",
			filters={
				"custom_base_reference": "Session Schedule",
				"custom_base_reference_name": ["in", names],
				"docstatus": ["!=", 2],
			},
			fields=["name", "custom_base_reference_name"],
			order_by="creation desc",
		):
			by_schedule.setdefault(row.custom_base_reference_name, row.name)

	for row in schedules:
		row["sales_order"] = by_schedule.get(row["name"]) or row.get("invoice_no")
	return schedules


def _patient_care_type_from_schedule(doc):
	if doc.get("admission_number"):
		return "IP"
	if doc.get("patient_visit"):
		return "OP"
	return None


def _patient_from_schedule_refs(admission_number=None, patient_visit=None):
	"""Resolve Patient from IP admission or OP visit context."""
	admission_number = (admission_number or "").strip() or None
	patient_visit = (patient_visit or "").strip() or None
	if patient_visit and frappe.db.exists("Patient Visit", patient_visit):
		return frappe.db.get_value("Patient Visit", patient_visit, "patient")
	if admission_number and frappe.db.exists("Inpatient Admission", admission_number):
		return frappe.db.get_value("Inpatient Admission", admission_number, "patient")
	return None


def _default_amount_from_template(session_type, amount=None, patient_care_type=None, patient=None):
	"""Template/OP rate, then TRICARE inclusive price + OP/IP discount when patient is insured.

	If the UI stamped the raw template rate, replace it with the insured charge.
	A manually typed amount that differs from the template rate is kept as an override.
	"""
	if not session_type or not frappe.db.exists("Healthcare Service Template", session_type):
		return flt(amount, 2)

	from healthcare.controllers.insurance_pricing import get_patient_insurance, resolve_charge
	from healthcare.healthcare.doctype.healthcare_service_template.healthcare_service_template import (
		get_healthcare_service_template_rate,
	)

	base = flt(
		get_healthcare_service_template_rate(
			template_name=session_type,
			patient_care_type=patient_care_type,
		),
		2,
	)
	submitted = flt(amount, 2)
	service_type = patient_care_type if patient_care_type in ("OP", "IP", "IOP") else None

	_patient_doc, insurance_doc = get_patient_insurance(patient)
	if insurance_doc:
		charged = resolve_charge(
			patient=patient,
			base_rate=base or submitted,
			patient_care_type=patient_care_type,
			template_dt="Healthcare Service Template",
			template_dn=session_type,
			service_type=service_type,
		)
		# Store Inclusive / catalog list price on Session Schedule.amount;
		# insurance % is applied when creating the Sales Order.
		list_rate = flt(charged["rate_before_discount"], 2)
		if submitted <= 0 or (base > 0 and submitted == base):
			return list_rate
		# Also replace if UI previously stamped the net insured charge.
		net_rate = flt(charged["rate"], 2)
		if net_rate > 0 and submitted == net_rate and list_rate != net_rate:
			return list_rate
		return submitted

	if submitted > 0:
		return submitted
	return base


@frappe.whitelist()
def get_session_schedule_amount(
	session_type: str | None = None,
	patient: str | None = None,
	patient_visit: str | None = None,
	admission_number: str | None = None,
	patient_care_type: str | None = None,
):
	"""Return the amount to show when picking a Healthcare Service Template on Session Schedule."""
	session_type = (session_type or "").strip()
	if not session_type:
		return {"amount": 0, "base_rate": 0, "discount_pct": 0, "insurance": None}

	patient = (patient or "").strip() or _patient_from_schedule_refs(admission_number, patient_visit)
	care_type = (patient_care_type or "").strip().upper() or None
	if not care_type:
		care_type = "IP" if (admission_number or "").strip() else "OP"

	from healthcare.controllers.insurance_pricing import resolve_charge
	from healthcare.healthcare.doctype.healthcare_service_template.healthcare_service_template import (
		get_healthcare_service_template_rate,
	)

	base = flt(
		get_healthcare_service_template_rate(
			template_name=session_type,
			patient_care_type=care_type,
		),
		2,
	)
	charged = resolve_charge(
		patient=patient,
		base_rate=base,
		patient_care_type=care_type,
		template_dt="Healthcare Service Template",
		template_dn=session_type,
		service_type=care_type if care_type in ("OP", "IP", "IOP") else None,
	)
	return {
		# Amount shown/stored = list (Inclusive price × multiplier); discount tracked on SO.
		"amount": flt(charged["rate_before_discount"], 2),
		"base_rate": flt(charged["base_rate"], 2),
		"discount_pct": flt(charged["discount_pct"], 2),
		"discount_amount": flt(charged.get("discount_amount") or 0, 2),
		"net_rate": flt(charged["rate"], 2),
		"insurance": charged.get("insurance"),
		"used_insurance_price": charged.get("used_insurance_price"),
		"patient": patient,
		"patient_care_type": care_type,
	}


def _resolve_session_schedule_billing(doc):
	template_name = doc.session_type
	if not template_name or not frappe.db.exists("Healthcare Service Template", template_name):
		frappe.throw(_("Healthcare Service Template is required"))

	template = frappe.get_doc("Healthcare Service Template", template_name)
	item_code = (template.item_code or "").strip()
	if not item_code or not frappe.db.exists("Item", item_code):
		frappe.throw(
			_("Healthcare Service Template {0} has no valid Item for billing.").format(template_name)
		)

	care_type = _patient_care_type_from_schedule(doc)
	patient = (
		doc.get("patient_num")
		or doc.get("patient")
		or _patient_from_schedule_refs(doc.get("admission_number"), doc.get("patient_visit"))
	)
	amount = _default_amount_from_template(
		template_name,
		doc.amount,
		patient_care_type=care_type,
		patient=patient,
	)

	if amount <= 0:
		frappe.throw(_("Amount is required to create a Sales Order."))

	description = doc.session_name or template.service_name or template_name
	return item_code, amount, description, patient


def _session_schedule_billing_refs(doc):
	"""Prefer admission, then visit, else the Session Schedule itself (standalone bill)."""
	admission = (doc.get("admission_number") or "").strip() or None
	visit = (doc.get("patient_visit") or "").strip() or None
	if admission:
		return "Inpatient Admission", admission
	if visit:
		return "Patient Visit", visit
	return "Session Schedule", doc.name


def _default_company(doc=None):
	"""Resolve company for session schedule / billing (single-company sites use default)."""
	if doc and doc.get("company"):
		return doc.company
	if doc and doc.get("admission_number"):
		company = frappe.db.get_value("Inpatient Admission", doc.admission_number, "company")
		if company:
			return company
	if doc and doc.get("patient_visit"):
		company = frappe.db.get_value("Patient Visit", doc.patient_visit, "company")
		if company:
			return company
	return frappe.defaults.get_user_default("company") or frappe.db.get_single_value(
		"Global Defaults", "default_company"
	)


def _create_session_schedule_sales_order(doc):
	existing = _existing_session_schedule_sales_order(doc.name)
	if existing:
		so = frappe.get_doc("Sales Order", existing)
		if cint(so.docstatus) == 0:
			so.flags.ignore_permissions = True
			so.submit()
		status = frappe.db.get_value("Session Schedule", doc.name, "transaction_status")
		return {"sales_order": so.name, "existing": True, "transaction_status": status}

	from healthcare.api.patient_file_no_charge import _ensure_patient_customer
	from healthcare.api.sales_order_cost_center import (
		apply_cost_center_to_sales_order,
		cost_center_from_visit_or_admission,
	)

	item_code, amount, description, patient = _resolve_session_schedule_billing(doc)
	patient = patient or (doc.get("patient_num") or "").strip() or None
	if not patient:
		frappe.throw(
			_("Patient is required to create a Sales Order. Link a patient, visit, or admission on the session.")
		)

	ref_type, ref_name = _session_schedule_billing_refs(doc)
	customer = _ensure_patient_customer(patient)

	company = _default_company(doc)
	if not company:
		frappe.throw(_("Company is required to create a Sales Order."))

	billing_date = getdate(doc.date or nowdate())

	from healthcare.controllers.insurance_pricing import (
		charge_list_and_discount,
		resolve_charge,
		sales_item_from_list_and_discount,
	)

	care_type = _patient_care_type_from_schedule(doc)
	charged = resolve_charge(
		patient=patient,
		base_rate=amount,
		patient_care_type=care_type,
		template_dt="Healthcare Service Template",
		template_dn=doc.session_type,
		service_type=care_type if care_type in ("OP", "IP", "IOP") else None,
	)
	parts = charge_list_and_discount(charged)
	# Prefer session amount as list when it matches insurance/template list; keep overrides.
	list_rate = flt(amount)
	insured_list = flt(parts["list_rate"])
	use_insurance_discount = insured_list > 0 and (
		list_rate <= 0 or abs(list_rate - insured_list) < 0.005
	)
	if use_insurance_discount:
		list_rate = insured_list

	so = frappe.new_doc("Sales Order")
	so.company = company
	so.customer = customer
	so.patient = patient
	if hasattr(so, "custom_patient"):
		so.custom_patient = patient

	patient_name = frappe.db.get_value("Patient", patient, "patient_name")
	if patient_name and hasattr(so, "custom_patient_name"):
		so.custom_patient_name = patient_name

	so.custom_reference_type = ref_type
	so.custom_reference_name = ref_name
	so.custom_base_reference = "Session Schedule"
	so.custom_base_reference_name = doc.name
	so.transaction_date = billing_date
	so.delivery_date = billing_date
	so.ignore_pricing_rule = 1

	so.append(
		"items",
		sales_item_from_list_and_discount(
			item_code=item_code,
			list_rate=list_rate,
			discount_pct=parts["discount_pct"] if use_insurance_discount else 0,
			net_rate=parts["net_rate"] if use_insurance_discount else list_rate,
			description=_("Session Schedule {0}: {1}").format(doc.name, description),
		),
	)

	cc = doc.cost_center or cost_center_from_visit_or_admission(ref_type, ref_name)
	apply_cost_center_to_sales_order(so, cc)
	so.insert(ignore_permissions=True)
	so.flags.ignore_permissions = True
	so.submit()

	doc.db_set("invoice_no", so.name, update_modified=False)
	if doc.transaction_status not in ("Cancelled", "Submitted"):
		doc.db_set("transaction_status", "Submitted", update_modified=False)

	return {"sales_order": so.name, "existing": False, "transaction_status": "Submitted"}


def _backfill_blank_patient_num():
	"""Fill patient_num from visit/admission when it was left blank (older create bug)."""
	frappe.db.sql(
		"""
		UPDATE `tabSession Schedule` ss
		INNER JOIN `tabPatient Visit` pv ON ss.patient_visit = pv.name
		SET ss.patient_num = pv.patient
		WHERE IFNULL(ss.patient_num, '') = ''
			AND IFNULL(ss.patient_visit, '') != ''
			AND IFNULL(pv.patient, '') != ''
		"""
	)
	frappe.db.sql(
		"""
		UPDATE `tabSession Schedule` ss
		INNER JOIN `tabInpatient Admission` ia ON ss.admission_number = ia.name
		SET ss.patient_num = ia.patient
		WHERE IFNULL(ss.patient_num, '') = ''
			AND IFNULL(ss.admission_number, '') != ''
			AND IFNULL(ia.patient, '') != ''
		"""
	)


def _session_schedule_list_fields():
	return [
		"name",
		"date",
		"admission_number",
		"patient_num",
		"patient_visit",
		"session_type",
		"session_name",
		"transaction_status",
		"company",
		"doctor",
		"doctor_name",
		"practitioner",
		"practitioner_name",
		"cost_center",
		"invoice_no",
		"doc_code",
		"doc_remarks",
		"visit_00_05",
		"sr_num",
		"from_time",
		"to_time",
		"amount",
	]


@frappe.whitelist()
def get_session_schedules(
	limit: int = 50,
	offset: int = 0,
	patient: str = None,
	admission_number: str = None,
	role_group: str = None,
	practitioner: str = None,
):
	"""Fetch session schedules with optional filtering by patient or admission number.

	role_group ('Doctor' / 'Nurse' / 'Consultant' / ...) limits rows to sessions whose
	assigned practitioner's Medical Role falls under that group — plus unassigned
	sessions, so nothing scheduled but not yet assigned disappears.

	practitioner filters by Session Schedule.practitioner (who entered the session).
	"""
	_backfill_blank_patient_num()

	filters = {}
	patient = (patient or "").strip() or None
	admission_number = (admission_number or "").strip() or None
	practitioner = (practitioner or "").strip() or None
	role_group = (role_group or "").strip() or None

	if admission_number:
		filters["admission_number"] = admission_number
	if practitioner:
		filters["practitioner"] = practitioner

	patient_or_filters = None
	if patient:
		visit_names = frappe.get_all(
			"Patient Visit",
			filters={"patient": patient},
			pluck="name",
		) or []
		admission_names = frappe.get_all(
			"Inpatient Admission",
			filters={"patient": patient},
			pluck="name",
		) or []
		patient_or_filters = [["patient_num", "=", patient]]
		if visit_names:
			patient_or_filters.append(["patient_visit", "in", visit_names])
		if admission_names:
			patient_or_filters.append(["admission_number", "in", admission_names])

	role_or_filters = None
	if role_group:
		from healthcare.api.common import get_medical_roles_under

		roles = list(get_medical_roles_under(role_group))
		practitioners = frappe.get_all(
			"Healthcare Practitioner",
			filters={"medical_role": ["in", roles]},
			pluck="name",
		) or ["__none__"]
		role_or_filters = [
			["doctor", "in", practitioners],
			["doctor", "is", "not set"],
		]

	fetch_limit = limit
	fetch_offset = offset
	or_filters = None

	if patient_or_filters and role_or_filters:
		# get_list can't AND two or_filter groups — over-fetch by patient, then role-filter.
		or_filters = patient_or_filters
		fetch_limit = max(limit * 10, 200)
		fetch_offset = 0
	elif patient_or_filters:
		or_filters = patient_or_filters
	elif role_or_filters:
		or_filters = role_or_filters

	schedules = frappe.get_list(
		"Session Schedule",
		filters=filters,
		or_filters=or_filters,
		fields=_session_schedule_list_fields(),
		limit_page_length=fetch_limit,
		limit_start=fetch_offset,
		order_by="date desc",
	)

	if patient_or_filters and role_or_filters:
		allowed_doctors = set()
		for row in role_or_filters:
			if row[0] == "doctor" and row[1] == "in":
				allowed_doctors.update(row[2] or [])
		schedules = [
			row
			for row in schedules
			if not row.get("doctor") or row.get("doctor") in allowed_doctors
		]
		total_count = len(schedules)
		schedules = schedules[offset : offset + limit]
	else:
		total_count = len(
			frappe.get_list(
				"Session Schedule",
				filters=filters,
				or_filters=or_filters,
				pluck="name",
			)
		)

	return {
		"data": _attach_sales_orders(schedules),
		"total_count": int(total_count or 0),
	}


@frappe.whitelist()
def create_session_schedule(data: dict):
	"""Create a new Session Schedule record."""
	if not data:
		frappe.throw(_("No data provided"))

	try:
		session_schedule = frappe.new_doc("Session Schedule")

		admission_number = (data.get("admission_number") or "").strip() or None
		patient_visit = (data.get("patient_visit") or "").strip() or None
		explicit_patient = (data.get("patient") or data.get("patient_num") or "").strip() or None
		care_type = "IP" if admission_number else ("OP" if patient_visit else None)

		session_schedule.date = data.get("date")
		session_schedule.session_type = data.get("session_type")
		session_schedule.session_name = data.get("session_name")
		session_schedule.doctor = data.get("doctor")
		practitioner = (data.get("practitioner") or "").strip()
		if practitioner:
			session_schedule.practitioner = practitioner
			session_schedule.practitioner_name = (
				data.get("practitioner_name")
				or frappe.db.get_value("Healthcare Practitioner", practitioner, "practitioner_name")
				or practitioner
			)
		session_schedule.cost_center = data.get("cost_center")
		session_schedule.from_time = data.get("from_time")
		session_schedule.to_time = data.get("to_time")
		session_schedule.admission_number = admission_number
		session_schedule.patient_visit = patient_visit
		session_schedule.company = _default_company(session_schedule)

		patient = (
			_patient_from_schedule_refs(admission_number, patient_visit)
			or explicit_patient
		)
		if patient:
			# patient_num is read_only on the form; set it in code so list filters work.
			session_schedule.patient_num = patient
		else:
			frappe.throw(
				_("Select a patient (or link a Patient Visit / Admission) before creating a session schedule.")
			)

		session_schedule.amount = _default_amount_from_template(
			data.get("session_type"),
			data.get("amount"),
			patient_care_type=care_type,
			patient=patient,
		) or None
		session_schedule.transaction_status = "Draft"

		session_schedule.insert(ignore_permissions=True)
		frappe.db.commit()

		row = session_schedule.as_dict()
		return _attach_sales_orders([row])[0]
	except Exception as e:
		frappe.log_error(frappe.get_traceback(), "create_session_schedule")
		frappe.throw(str(e))


@frappe.whitelist()
def create_sales_order_from_session_schedule(session_schedule_name: str):
	"""Bill a Session Schedule (create or return linked Sales Order)."""
	if not session_schedule_name:
		frappe.throw(_("Session Schedule name is required"))

	if not frappe.db.exists("Session Schedule", session_schedule_name):
		frappe.throw(_("Session Schedule {0} not found").format(session_schedule_name))

	doc = frappe.get_doc("Session Schedule", session_schedule_name)
	result = _create_session_schedule_sales_order(doc)
	frappe.db.commit()
	row = doc.as_dict()
	result.update(_attach_sales_orders([row])[0])
	return result


@frappe.whitelist()
def update_session_schedule_status(session_schedule_name: str, status: str):
	"""Update the status of a Session Schedule."""
	assert_editing_allowed()
	if not session_schedule_name or not status:
		frappe.throw(_("Session Schedule name and status are required"))

	status = (status or "").strip()
	if status not in SESSION_SCHEDULE_STATUSES:
		frappe.throw(_("Invalid status: {0}").format(status))

	session_schedule = frappe.get_doc("Session Schedule", session_schedule_name)
	if session_schedule.transaction_status == "Cancelled" and status != "Cancelled":
		frappe.throw(_("Cancelled session schedules cannot be reactivated."))

	session_schedule.transaction_status = status
	session_schedule.save(ignore_permissions=True)
	frappe.db.commit()

	row = session_schedule.as_dict()
	return _attach_sales_orders([row])[0]
